File: Wawaji-Client-iOS/Wawaji-Client-iOS/ZinianWawajiController.m
//
//  ZinianWawajiController.m
//  Wawaji-Client-iOS
//
//  Created by suleyu on 2017/12/2.
//  Copyright © 2017 Agora. All rights reserved.
//

#import "ZinianWawajiController.h"
#import "SocketRocket.h"

static NSString * const kWebSocketUrl = @"Your Wawaji Controlling WebSocket Url";

@interface ZinianWawajiController () <SRWebSocketDelegate>
{
    SRWebSocket *webSocket;
}
@end

@implementation ZinianWawajiController

- (void)dealloc {
    [webSocket close];
    webSocket.delegate = nil;
    webSocket = nil;
}

- (void)initialize {
    [self connectWebSocket];
}

- (void)insertCion {
    NSDictionary *msgDic = @{@"type" : @"Insert", @"data" : @"", @"extra" : @(123456)};
    [self sendWebSocketMessage:msgDic];
}

- (BOOL)startMoveUp {
    NSDictionary *msgDic = @{@"type" : @"Control", @"data" : @"u"};
    [self sendWebSocketMessage:msgDic];
    return YES;
}

- (BOOL)stopMoveUp {
    return NO;
}

- (BOOL)startMoveDown {
    NSDictionary *msgDic = @{@"type" : @"Control", @"data" : @"d"};
    [self sendWebSocketMessage:msgDic];
    return YES;
}

- (BOOL)stopMoveDown {
    return NO;
}

- (BOOL)startMoveLeft {
    NSDictionary *msgDic = @{@"type" : @"Control", @"data" : @"l"};
    [self sendWebSocketMessage:msgDic];
    return YES;
}

- (BOOL)stopMoveLeft {
    return NO;
}

- (BOOL)startMoveRight {
    NSDictionary *msgDic = @{@"type" : @"Control", @"data" : @"r"};
    [self sendWebSocketMessage:msgDic];
    return YES;
}

- (BOOL)stopMoveRight {
    return NO;
}

- (BOOL)fetch {
    NSDictionary *msgDic = @{@"type" : @"Control", @"data" : @"b"};
    [self sendWebSocketMessage:msgDic];
    return YES;
}

// MARK: - WebSocket

- (void)connectWebSocket {
    [webSocket close];
    webSocket.delegate = nil;
    webSocket = nil;
    
    SRWebSocket *newWebSocket = [[SRWebSocket alloc] initWithURL:[NSURL URLWithString:kWebSocketUrl]];
    newWebSocket.delegate = self;
    [newWebSocket open];
}

- (void)sendWebSocketMessage:(NSDictionary *)message {
    NSData *data = [NSJSONSerialization dataWithJSONObject:message options:NSJSONWritingPrettyPrinted error:nil];
    NSString *string = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    NSLog(@"sendWebSocketMessage: %@", string);
    [webSocket send:string];
}

- (void)webSocketDidOpen:(SRWebSocket *)newWebSocket {
    webSocket = newWebSocket;
}

- (void)webSocket:(SRWebSocket *)webSocket didFailWithError:(NSError *)error {
    [self connectWebSocket];
}

- (void)webSocket:(SRWebSocket *)webSocket didCloseWithCode:(NSInteger)code reason:(NSString *)reason wasClean:(BOOL)wasClean {
    [self connectWebSocket];
}

- (void)webSocket:(SRWebSocket *)webSocket didReceiveMessage:(id)message {
}

@end

File: Wawaji-Client-iOS-Swift/Wawaji-Client/ZinianWawajiController.swift
//
//  ZinianWawajiController.swift
//  Wawaji-Client
//
//  Created by suleyu on 2017/12/1.
//  Copyright © 2017 Agora. All rights reserved.
//

import Foundation
import SocketRocket

class ZinianWawajiController: NSObject, WawajiController {
    let webSocketUrl: String = "Your Wawaji Controlling WebSocket Url"
    
    var webSocket : SRWebSocket?
    var fetchResult : ((Bool) -> Void)?
    
    deinit {
        webSocket?.close()
        webSocket?.delegate = nil
        webSocket = nil
    }
    
    func initialize() {
        connectWebSocket()
    }
    
    func insertCion() {
        let messageDic : [String : Any] = ["type" : "Insert", "data" : "", "extra" : 123456]
        sendWebSocketMessage(messageDic)
    }
    
    func startMoveUp() -> Bool {
        let messageDic : [String : Any] = ["type" : "Control", "data" : "u"]
        sendWebSocketMessage(messageDic)
        return true
    }
    
    func stopMoveUp() -> Bool {
        return false
    }
    
    func startMoveDown() -> Bool {
        let messageDic : [String : Any] = ["type" : "Control", "data" : "d"]
        sendWebSocketMessage(messageDic)
        return true
    }
    
    func stopMoveDown() -> Bool {
        return false
    }
    
    func startMoveLeft() -> Bool {
        let messageDic : [String : Any] = ["type" : "Control", "data" : "l"]
        sendWebSocketMessage(messageDic)
        return true
    }
    
    func stopMoveLeft() -> Bool {
        return false
    }
    
    func startMoveRight() -> Bool {
        let messageDic : [String : Any] = ["type" : "Control", "data" : "r"]
        sendWebSocketMessage(messageDic)
        return true
    }
    
    func stopMoveRight() -> Bool {
        return false
    }
    
    func fetch() -> Bool {
        let messageDic : [String : Any] = ["type" : "Control", "data" : "b"]
        sendWebSocketMessage(messageDic)
        return true
    }
    
    private func connectWebSocket() {
        if let oldWebSocket = webSocket {
            oldWebSocket.close()
            oldWebSocket.delegate = nil
            webSocket = nil
        }
        
        let newWebSocket = SRWebSocket(url: URL(string: webSocketUrl))
        newWebSocket!.delegate = self
        newWebSocket!.open()
    }
    
    private func sendWebSocketMessage(_ message: [String: Any]) {
        let data = try! JSONSerialization.data(withJSONObject: message, options: .prettyPrinted)
        let string = NSString(data: data, encoding: String.Encoding.utf8.rawValue)! as String
        print("sendWebSocketMessage: \(string)");
        webSocket!.send(string)
    }
}

extension ZinianWawajiController : SRWebSocketDelegate {
    func webSocketDidOpen(_ newWebSocket: SRWebSocket!) {
        webSocket = newWebSocket
    }
    
    func webSocket(_ webSocket: SRWebSocket!, didFailWithError error: Error!) {
        connectWebSocket()
    }
    
    func webSocket(_ webSocket: SRWebSocket!, didCloseWithCode code: Int, reason: String!, wasClean: Bool) {
        connectWebSocket()
    }
    
    func webSocket(_ webSocket: SRWebSocket!, didReceiveMessage message: Any!) {
    }
}
